runners/jet/src/main/java/org/apache/beam/runners/jet/processors/AbstractParDoP.java

```diff
@@ -0,0 +1,502 @@
+import org.apache.beam.runners.core.SideInputHandler;
+import org.apache.beam.runners.core.SideInputReader;
+import org.apache.beam.runners.core.construction.SerializablePipelineOptions;
+import org.apache.beam.runners.jet.DAGBuilder;
+import org.apache.beam.runners.jet.Utils;
+import org.apache.beam.runners.jet.metrics.JetMetricsContainer;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.metrics.MetricsEnvironment;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.DoFnSchemaInformation;
+import org.apache.beam.sdk.transforms.reflect.DoFnInvoker;
+import org.apache.beam.sdk.transforms.reflect.DoFnInvokers;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.WindowingStrategy;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.Lists;
+
+abstract class AbstractParDoP<InputT, OutputT> implements Processor {
+
+  private final SerializablePipelineOptions pipelineOptions;
+  private final DoFn<InputT, OutputT> doFn;
+  private final WindowingStrategy<?, ?> windowingStrategy;
+  private final DoFnSchemaInformation doFnSchemaInformation;
+  private final Map<TupleTag<?>, int[]> outputCollToOrdinals;
+  private final TupleTag<OutputT> mainOutputTag;
+  private final Coder<InputT> inputCoder;
+  private final Map<PCollectionView<?>, Coder<?>> sideInputCoders;
+  private final Map<TupleTag<?>, Coder<?>> outputCoders;
+  private final Coder<InputT> inputValueCoder;
+  private final Map<TupleTag<?>, Coder<?>> outputValueCoders;
+  private final Map<Integer, PCollectionView<?>> ordinalToSideInput;
+  private final String ownerId; // do not remove, useful for debugging
+  private final String stepId;
+
+  DoFnRunner<InputT, OutputT> doFnRunner;
+  JetOutputManager outputManager;
+  private DoFnInvoker<InputT, OutputT> doFnInvoker;
+  private SideInputHandler sideInputHandler;
+  private JetMetricsContainer metricsContainer;
+  private SimpleInbox bufferedItems;
+  private Set<Integer> completedSideInputs = new HashSet<>();
+  private SideInputReader sideInputReader;
+  private Outbox outbox;
+
+  AbstractParDoP(
+      DoFn<InputT, OutputT> doFn,
+      WindowingStrategy<?, ?> windowingStrategy,
+      DoFnSchemaInformation doFnSchemaInformation,
+      Map<TupleTag<?>, int[]> outputCollToOrdinals,
+      SerializablePipelineOptions pipelineOptions,
+      TupleTag<OutputT> mainOutputTag,
+      Coder<InputT> inputCoder,
+      Map<PCollectionView<?>, Coder<?>> sideInputCoders,
+      Map<TupleTag<?>, Coder<?>> outputCoders,
+      Coder<InputT> inputValueCoder,
+      Map<TupleTag<?>, Coder<?>> outputValueCoders,
+      Map<Integer, PCollectionView<?>> ordinalToSideInput,
+      String ownerId,
+      String stepId) {
+    this.pipelineOptions = pipelineOptions;
+    this.doFn = Utils.serde(doFn);
```

Comment:
Why is this required? Maybe for local execution? Otherwise I'd assume that this would be cloned anyways when shipped to the cluster.

Reply:
Yes, for local execution. This is cloned when shipped to each member of a cluster, but then, inside one member there can be multiple processors using it. Leaving it as is.